core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -159,18 +160,49 @@ public void commit(TableMetadata base, TableMetadata metadata) {
     int nextVersion = (current.first() != null ? current.first() : 0) + 1;
     Path finalMetadataFile = metadataFilePath(nextVersion, codec);
     FileSystem fs = getFileSystem(tempMetadataFile, conf);
-
-    // this rename operation is the atomic commit operation
-    renameToFinal(fs, tempMetadataFile, finalMetadataFile, nextVersion);
-
-    LOG.info("Committed a new metadata file {}", finalMetadataFile);
-
-    // update the best-effort version pointer
-    writeVersionHint(nextVersion);
-
-    deleteRemovedMetadataFiles(base, metadata);
-
-    this.shouldRefresh = true;
+    boolean versionCommitSuccess = false;
+    boolean useObjectStore =
+        metadata.propertyAsBoolean(
+            TableProperties.OBJECT_STORE_ENABLED, TableProperties.OBJECT_STORE_ENABLED_DEFAULT);
+    try {
+      if (!lockManager.acquire(finalMetadataFile.toString(), tempMetadataFile.toString())) {
+        throw new CommitFailedException(
+            "Failed to acquire lock on file: %s with owner: %s",
+            finalMetadataFile, tempMetadataFile);
+      }
+      versionCommitSuccess =
+          commitNewVersion(fs, tempMetadataFile, finalMetadataFile, nextVersion, useObjectStore);
+      if (!versionCommitSuccess) {
+        throw new CommitFailedException(
+            "Can not commit newMetaData because version [%s] has already been committed. commitVersion=[%s],tempMetaData=[%s],finalMetaData=[%s].Are there other clients running in parallel with the current task?",
+            nextVersion, nextVersion, tempMetadataFile, finalMetadataFile);
+      }
+      this.shouldRefresh = true;
+      LOG.info("Committed a new metadata file {}", finalMetadataFile);
+      // update the best-effort version pointer
+      writeVersionHint(fs, nextVersion);
+      deleteRemovedMetadataFiles(base, metadata);
+    } catch (CommitStateUnknownException e) {
+      this.shouldRefresh = true;
+      throw e;
+    } catch (Throwable e) {
+      this.shouldRefresh = versionCommitSuccess;
+      if (!versionCommitSuccess) {
+        io().deleteFile(tempMetadataFile.toString());
+        throw new CommitFailedException(e);
+      }
+    } finally {
+      try{
+        if (!lockManager.release(finalMetadataFile.toString(), tempMetadataFile.toString())) {
```

Comment:
Although lockManager.release can't throw any exceptions, I don't really trust it.